src/kakarot/precompiles/sha256.cairo

```diff
@@ -0,0 +1,455 @@
+
+%lang starknet
+
+// Starkware dependencies
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.registers import get_fp_and_pc
+from starkware.cairo.common.cairo_builtins import HashBuiltin, BitwiseBuiltin
+from starkware.cairo.common.math import assert_nn_le, unsigned_div_rem
+from starkware.cairo.common.math_cmp import is_le_felt
+from starkware.cairo.common.memcpy import memcpy
+from starkware.cairo.common.memset import memset
+from starkware.cairo.common.pow import pow
+
+// Internal dependencies
+from utils.sha_256.packed_sha256 import (
+    BLOCK_SIZE,
+    compute_message_schedule,
+    sha2_compress,
+    get_round_constants,
+)
+from utils.utils import Helpers
+
+// @title SHA2-256 Precompile related functions.
+// @notice This file contains the logic required to run the SHA2-256 precompile
+// @author @ftupas
+// @custom:namespace PrecompileSHA256
+namespace PrecompileSHA256 {
+    const PRECOMPILE_ADDRESS = 0x02;
+    const GAS_COST_SHA256 = 60;
+
+    // @notice Run the precompile.
+    // @param input_len The length of input array.
+    // @param input The input array.
+    // @return The output length, output array, and gas usage of precompile.
+    func run{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(_address: felt, input_len: felt, input: felt*) -> (
+        output_len: felt, output: felt*, gas_used: felt
+    ) {
+        alloc_locals;
+
+        // Copy input array
+        let (arr: felt*) = alloc();
+        memcpy(arr, input, input_len);
+
+        // Zero-pad bytes array
+        // ex. 'rld\x00'
+        let (q: felt, r: felt) = unsigned_div_rem(input_len, 4);
+        if (r != 0) {
+            // Append zero elements at the end of array
+            Helpers.fill(4 - r, arr + input_len, 0);
+        }
+        let arr_len = (q + 1) * 4;
+
+        // Prepare input bytes array to words of 32 bits (big endian).
+        let (prepared_input: felt*) = alloc();
+        let (prepared_input_len, prepared_input: felt*) = Helpers.bytes_to_words_32bit_array(
+            arr_len, arr, 0, prepared_input
+        );
+
+        // Compute hash, use input_len for number of bytes
+        let (local sha256_ptr: felt*) = alloc();
+        let sha256_ptr_start = sha256_ptr;
+        let (hash) = SHA256.sha256{sha256_ptr=sha256_ptr}(prepared_input, input_len);
+
+        // Finalize hash
+        SHA256.finalize_sha256(sha256_ptr_start=sha256_ptr_start, sha256_ptr_end=sha256_ptr);
+
+        // Split words and return bytes hash code.
+        let (hash_bytes_array: felt*) = alloc();
+        let (_, hash_bytes_array: felt*) = Helpers.words_32bit_to_bytes_array(
+            8, hash, 0, hash_bytes_array
+        );
+        let (minimum_word_size) = Helpers.minimum_word_count(input_len);
+        return (32, hash_bytes_array, 3 * minimum_word_size + GAS_COST_SHA256);
```

Comment:
Entrypoint here. This is the main implementation for the precompile. Original code from [Cartridge's](https://github.com/cartridge-gg/cairo-sha256) which is based on Starkware's implementation + testing. It takes in 32-bit words (BE) so we have to prepare the bytes array input to conform to this format and zero-pad. The output is 8 32-bit words so we need to transform them into back into bytes array